src/silly_sprites/opengl/shader_program.cpp

```diff
@@ -0,0 +1,179 @@
+        auto compile_single = [&](GLuint& id, char const* single_source) -> GLint {
+            id = glCreateShader(sly::to_underlying(type));
+            glShaderSource(id, 1, &single_source, nullptr);
+            glCompileShader(id);
+            auto success = GLint{};
+            glGetShaderiv(id, GL_COMPILE_STATUS, &success);
+            return success;
+        };
+
+        auto id = GLuint{};
+        auto success = compile_single(id, source.data());
+
+        if (not success) {
+
+            auto message = error_message(id);
+            spdlog::critical("ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}", get_name_from_type(type), message);
+            glDeleteShader(id);
+
+            spdlog::info("SHADER::{}::TRY_FALLBACK_COMPILATION", get_name_from_type(type));
+            switch (type) {
+                case Type::Vertex:
+                    success = compile_single(id, fallback_vertex_source);
+                    if (not success) {
+                        auto message2 = error_message(id);
+                        spdlog::critical(
+                                "ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}",
+                                get_name_from_type(type),
+                                message2
+                        );
+                        glDeleteShader(id);
+                        throw GLError(
+                                GLErrorType::FailedToCompileVertexShader,
+                                fmt::format("first error: {}\nsecond error: {}", message, message2)
+                        );
+                    }
+                    break;
+
+                case Type::Geometry:
+                    spdlog::critical("ERROR::SHADER::{}::NO_FALLBACK", get_name_from_type(type));
+                    throw GLError(GLErrorType::FailedToCompileGeometryShader, message);
+                    break;
+
+                case Type::Fragment:
+                    success = compile_single(id, fallback_fragment_source);
+                    if (not success) {
+                        auto message2 = error_message(id);
+                        spdlog::critical(
+                                "ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}",
+                                get_name_from_type(type),
+                                message2
+                        );
+                        glDeleteShader(id);
+                        throw GLError(
+                                GLErrorType::FailedToCompileVertexShader,
+                                fmt::format("first error: {}\nsecond error: {}", message, message2)
+                        );
+                    }
+                    break;
+            }
+        }
+
+        spdlog::info("SUCCESS::SHADER::{}::COMPILATION", get_name_from_type(type));
+        return id;
+    }
+
+    void ShaderProgram::attach_shader(GLuint const shader) const {
+        glAttachShader(m_program_name, shader);
+        glDeleteShader(shader);
+    }
```

Comment:
I don't really like this function since it just assumes that `m_program_name` has a valid value before entering the function. Maybe make this function `static` and take `program_name` as parameter (to avoid accessing the object's state).